File: server/src/services/exportService.ts
import PDFDocument from 'pdfkit';
import {
  Document,
  Paragraph,
  TextRun,
  Table,
  TableRow,
  TableCell,
  HeadingLevel,
  AlignmentType,
  BorderStyle,
  ShadingType,
  PageNumber,
  Header,
  Footer,
  Packer,
  WidthType,
} from 'docx';
import axios from 'axios';
import { Response } from 'express';
import type { Message } from '../../../shared/types';

interface ExportData {
  videoId: string;
  title: string;
  channelName: string;
  duration: string;
  thumbnailUrl: string;
  summary: string;
  chatHistory: Message[];
  includeTranscript: boolean;
  transcript?: string;
}

/**
 * Fetches a YouTube thumbnail and returns it as a Buffer.
 */
async function fetchThumbnail(thumbnailUrl: string): Promise<Buffer | null> {
  try {
    const response = await axios.get(thumbnailUrl, {
      responseType: 'arraybuffer',
      timeout: 8000,
    });
    return Buffer.from(response.data);
  } catch {
    // Try fallback thumbnail URL
    try {
      const videoIdMatch = thumbnailUrl.match(/\/vi\/([^/]+)\//);
      if (videoIdMatch) {
        const fallbackUrl = `https://img.youtube.com/vi/${videoIdMatch[1]}/hqdefault.jpg`;
        const fallback = await axios.get(fallbackUrl, { responseType: 'arraybuffer', timeout: 5000 });
        return Buffer.from(fallback.data);
      }
    } catch {
      // ignore
    }
    return null;
  }
}

function formatTimestamp(isoTimestamp: string): string {
  try {
    return new Date(isoTimestamp).toLocaleString();
  } catch {
    return isoTimestamp;
  }
}

/**
 * Generates a PDF export and pipes it to the response.
 */
export async function generatePdf(data: ExportData, res: Response): Promise<void> {
  const doc = new PDFDocument({ margin: 50, size: 'A4' });

  res.setHeader('Content-Type', 'application/pdf');
  res.setHeader('Content-Disposition', `attachment; filename="chat-export-${data.videoId}.pdf"`);
  doc.pipe(res);

  const thumbnailBuffer = await fetchThumbnail(data.thumbnailUrl);

  // Colours
  const USER_BG = '#E3F2FD';
  const AI_BG = '#F5F5F5';
  const ACCENT = '#1565C0';
  const HEADER_BG = '#1A237E';

  // ── Cover Page ──────────────────────────────────────────────
  doc.rect(0, 0, doc.page.width, 160).fill(HEADER_BG);
  doc.fillColor('white').fontSize(22).font('Helvetica-Bold').text('YouTube AI Chat Export', 50, 50, {
    width: doc.page.width - 100,
  });
  doc.fontSize(12).font('Helvetica').text(`Exported on ${new Date().toLocaleDateString()}`, 50, 82);

  let yPos = 175;

  // Thumbnail
  if (thumbnailBuffer) {
    try {
      doc.image(thumbnailBuffer, 50, yPos, { width: 200, height: 112 });
    } catch {
      // Skip image if it can't be rendered
    }
  }

  // Video metadata
  doc.fillColor('#000000');
  const metaX = thumbnailBuffer ? 270 : 50;
  doc.fontSize(16).font('Helvetica-Bold').fillColor(ACCENT).text(data.title, metaX, yPos, { width: doc.page.width - metaX - 50 });
  yPos += 24;
  doc.fontSize(11).font('Helvetica').fillColor('#555555')
    .text(`Channel: ${data.channelName}`, metaX, yPos)
    .text(`Duration: ${data.duration}`, metaX, yPos + 16)
    .text(`Video ID: ${data.videoId}`, metaX, yPos + 32);

  yPos = thumbnailBuffer ? yPos + 90 : yPos + 60;

  // ── Summary Section ────────────────────────────────────────
  doc.moveTo(50, yPos).lineTo(doc.page.width - 50, yPos).stroke('#CCCCCC');
  yPos += 15;

  doc.fontSize(14).font('Helvetica-Bold').fillColor(ACCENT).text('Video Summary', 50, yPos);
  yPos += 22;
  doc.fontSize(10).font('Helvetica').fillColor('#333333').text(data.summary, 50, yPos, {
    width: doc.page.width - 100,
  });
  yPos = doc.y + 20;

  // ── Conversation ────────────────────────────────────────────
  doc.moveTo(50, yPos).lineTo(doc.page.width - 50, yPos).stroke('#CCCCCC');
  yPos += 15;
  doc.fontSize(14).font('Helvetica-Bold').fillColor(ACCENT).text('Conversation', 50, yPos);
  yPos += 25;

  for (const msg of data.chatHistory) {
    const isUser = msg.role === 'user';
    const bgColor = isUser ? USER_BG : AI_BG;
    const label = isUser ? 'You' : 'AI Assistant';
    const labelColor = isUser ? '#1565C0' : '#2E7D32';

    // Check if we need a new page
    if (yPos > doc.page.height - 150) {
      doc.addPage();
      yPos = 50;
    }

    // Message bubble
    const msgWidth = doc.page.width - 100;
    const msgText = msg.content;
    const textHeight = doc.heightOfString(msgText, { width: msgWidth - 20 });
    const bubbleHeight = textHeight + 35;

    doc.roundedRect(50, yPos, msgWidth, bubbleHeight, 6).fill(bgColor);
    doc.fontSize(9).font('Helvetica-Bold').fillColor(labelColor).text(label, 62, yPos + 8);
    doc.fontSize(9).font('Helvetica').fillColor('#888888').text(formatTimestamp(msg.timestamp), msgWidth - 50, yPos + 8, { align: 'right', width: 150 });
    doc.fontSize(10).font('Helvetica').fillColor('#222222').text(msgText, 62, yPos + 22, { width: msgWidth - 20 });

    yPos = doc.y + 12;
  }

  // ── Optional Transcript Appendix ───────────────────────────
  if (data.includeTranscript && data.transcript) {
    doc.addPage();
    doc.fontSize(16).font('Helvetica-Bold').fillColor(ACCENT).text('Full Transcript', 50, 50);
    doc.fontSize(9).font('Helvetica').fillColor('#333333').text(data.transcript, 50, 80, {
      width: doc.page.width - 100,
      columns: 2,
      columnGap: 20,
    });
  }

  // Add page numbers
  const range = doc.bufferedPageRange();
  for (let i = range.start; i < range.start + range.count; i++) {
    doc.switchToPage(i);
    doc.fontSize(9).fillColor('#888888').text(
      `${data.title} — Page ${i + 1} of ${range.count}`,
      50,
      doc.page.height - 30,
      { width: doc.page.width - 100, align: 'center' }
    );
  }

  doc.end();
}

/**
 * Generates a DOCX export and sends it as a response.
 */
export async function generateDocx(data: ExportData, res: Response): Promise<void> {
  const sections = [];

  // ── Cover / Header ─────────────────────────────────────────
  const titleParagraph = new Paragraph({
    text: data.title,
    heading: HeadingLevel.HEADING_1,
    spacing: { after: 200 },
  });

  const metaParagraphs = [
    new Paragraph({ text: `Channel: ${data.channelName}`, spacing: { after: 100 } }),
    new Paragraph({ text: `Duration: ${data.duration}`, spacing: { after: 100 } }),
    new Paragraph({ text: `Exported on: ${new Date().toLocaleDateString()}`, spacing: { after: 300 } }),
  ];

  // ── Summary ─────────────────────────────────────────────────
  const summarySection = [
    new Paragraph({
      text: 'Video Summary',
      heading: HeadingLevel.HEADING_2,
      spacing: { before: 300, after: 150 },
    }),
    new Paragraph({
      text: data.summary,
      spacing: { after: 300 },
    }),
  ];

  // ── Conversation ─────────────────────────────────────────────
  const conversationHeader = new Paragraph({
    text: 'Conversation',
    heading: HeadingLevel.HEADING_2,
    spacing: { before: 300, after: 150 },
  });

  const messageElements: (Paragraph | Table)[] = [];

  for (const msg of data.chatHistory) {
    const isUser = msg.role === 'user';
    const label = isUser ? 'You' : 'AI Assistant';

    if (isUser) {
      // User messages in shaded table cells
      messageElements.push(
        new Table({
          width: { size: 100, type: WidthType.PERCENTAGE },
          rows: [
            new TableRow({
              children: [
                new TableCell({
                  shading: { type: ShadingType.CLEAR, color: 'E3F2FD', fill: 'E3F2FD' },
                  borders: {
                    top: { style: BorderStyle.NONE },
                    bottom: { style: BorderStyle.NONE },
                    left: { style: BorderStyle.THICK, color: '1565C0', size: 6 },
                    right: { style: BorderStyle.NONE },
                  },
                  children: [
                    new Paragraph({
                      children: [new TextRun({ text: label, bold: true, color: '1565C0', size: 18 })],
                      spacing: { after: 80 },
                    }),
                    new Paragraph({ text: msg.content, spacing: { after: 80 } }),
                    new Paragraph({
                      children: [new TextRun({ text: formatTimestamp(msg.timestamp), color: '888888', size: 16 })],
                    }),
                  ],
                }),
              ],
            }),
          ],
        })
      );
    } else {
      // AI messages as paragraphs with left border
      messageElements.push(
        new Paragraph({
          children: [new TextRun({ text: label, bold: true, color: '2E7D32', size: 18 })],
          spacing: { before: 200, after: 80 },
          border: { left: { style: BorderStyle.THICK, color: '2E7D32', size: 6 } },
          indent: { left: 180 },
        }),
        new Paragraph({
          text: msg.content,
          spacing: { after: 80 },
          border: { left: { style: BorderStyle.THICK, color: '2E7D32', size: 6 } },
          indent: { left: 180 },
        }),
        new Paragraph({
          children: [new TextRun({ text: formatTimestamp(msg.timestamp), color: '888888', size: 16 })],
          border: { left: { style: BorderStyle.THICK, color: '2E7D32', size: 6 } },
          indent: { left: 180 },
          spacing: { after: 200 },
        })
      );
    }

    // Spacer paragraph
    messageElements.push(new Paragraph({ text: '', spacing: { after: 100 } }));
  }

  // ── Transcript Appendix ──────────────────────────────────────
  const appendixParagraphs: Paragraph[] = [];
  if (data.includeTranscript && data.transcript) {
    appendixParagraphs.push(
      new Paragraph({
        text: 'Full Transcript',
        heading: HeadingLevel.HEADING_2,
        pageBreakBefore: true,
        spacing: { after: 200 },
      }),
      new Paragraph({
        text: data.transcript,
        spacing: { after: 200 },
        style: 'Normal',
      })
    );
  }

  const doc = new Document({
    title: data.title,
    subject: `Chat export for YouTube video: ${data.videoId}`,
    creator: 'YouTube AI Chat Agent',
    sections: [
      {
        headers: {
          default: new Header({
            children: [
              new Paragraph({
                children: [
                  new TextRun({ text: data.title, bold: true, size: 16 }),
                  new TextRun({ text: ' — YouTube AI Chat Export', size: 16 }),
                ],
                alignment: AlignmentType.RIGHT,
              }),
            ],
          }),
        },
        footers: {
          default: new Footer({
            children: [
              new Paragraph({
                children: [
                  new TextRun({ text: 'Page ', size: 16 }),
                  new TextRun({
                    children: [PageNumber.CURRENT],
                    size: 16,
                  }),
                  new TextRun({ text: ' of ', size: 16 }),
                  new TextRun({
                    children: [PageNumber.TOTAL_PAGES],
                    size: 16,
                  }),
                ],
                alignment: AlignmentType.CENTER,
              }),
            ],
          }),
        },
        children: [
          titleParagraph,
          ...metaParagraphs,
          ...summarySection,
          conversationHeader,
          ...messageElements,
          ...appendixParagraphs,
        ],
      },
    ],
  });

  const buffer = await Packer.toBuffer(doc);

  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document');
  res.setHeader('Content-Disposition', `attachment; filename="chat-export-${data.videoId}.docx"`);
  res.send(buffer);
}

File: server-python/app/services/export_service.py
import io
import os
import httpx
from datetime import datetime
from typing import Any
from docx import Document
from docx.shared import (
    Inches,
    Pt,
    RGBColor,
)
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.enum.style import WD_STYLE_TYPE
from docx.oxml.ns import qn
from docx.oxml import OxmlElement
from fpdf import FPDF

from app.models import Message


class ExportData:
    def __init__(
        self,
        video_id: str,
        title: str,
        channel_name: str,
        duration: str,
        thumbnail_url: str,
        summary: str,
        chat_history: list[Message],
        include_transcript: bool = False,
        transcript: str | None = None,
    ):
        self.video_id = video_id
        self.title = title
        self.channel_name = channel_name
        self.duration = duration
        self.thumbnail_url = thumbnail_url
        self.summary = summary
        self.chat_history = chat_history
        self.include_transcript = include_transcript
        self.transcript = transcript


async def fetch_thumbnail(thumbnail_url: str) -> bytes | None:
    """Fetches a YouTube thumbnail and returns it as bytes."""
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(thumbnail_url, timeout=8.0)
            response.raise_for_status()
            return response.content
    except Exception:
        # Try fallback URL
        try:
            video_id_match = thumbnail_url.rsplit("/vi/", 1)[-1].split("/")[0]
            if video_id_match:
                fallback_url = (
                    f"https://img.youtube.com/vi/{video_id_match}/hqdefault.jpg"
                )
                async with httpx.AsyncClient() as client:
                    response = await client.get(fallback_url, timeout=5.0)
                    response.raise_for_status()
                    return response.content
        except Exception:
            pass
    return None


def format_timestamp(iso_timestamp: str) -> str:
    """Formats ISO timestamp to readable string."""
    try:
        return datetime.fromisoformat(iso_timestamp.replace("Z", "+00:00")).strftime(
            "%Y-%m-%d %H:%M:%S"
        )
    except Exception:
        return iso_timestamp


async def generate_pdf(data: ExportData) -> bytes:
    """Generates a PDF export and returns it as bytes."""
    pdf = PDF()
    pdf.add_page()

    # Colors
    USER_BG = (227, 242, 253)  # #E3F2FD
    AI_BG = (245, 245, 245)  # #F5F5F5
    ACCENT = (21, 101, 192)  # #1565C0
    HEADER_BG = (26, 35, 126)  # #1A237E

    # Cover section - header bar
    pdf.set_fill_color(*HEADER_BG)
    pdf.rect(0, 0, 210, 40, "F")

    pdf.set_text_color(255, 255, 255)
    pdf.set_font("helvetica", "B", 18)
    pdf.cell(0, 25, "YouTube AI Chat Export", ln=True, align="C")

    pdf.set_font("helvetica", "", 10)
    pdf.ln(5)
    pdf.cell(
        0, 10, f"Exported on {datetime.now().strftime('%Y-%m-%d')}", ln=True, align="C"
    )

    pdf.ln(15)

    # Thumbnail and metadata
    thumbnail_bytes = await fetch_thumbnail(data.thumbnail_url)
    x_start = 10
    y_start = pdf.get_y()

    if thumbnail_bytes:
        tmp_path = None
        try:
            # Save to temp file for FPDF
            import tempfile

            with tempfile.NamedTemporaryFile(suffix=".jpg", delete=False) as tmp:
                tmp.write(thumbnail_bytes)
                tmp_path = tmp.name

            pdf.image(tmp_path, x=x_start, y=y_start, w=60, h=34)
            meta_x = 75
        except Exception:
            meta_x = x_start
        finally:
            if tmp_path:
                try:
                    os.unlink(tmp_path)
                except Exception:
                    pass
    else:
        meta_x = x_start

    # Write metadata aligned next to thumbnail (FPDF2 doesn't support `x=` on cell)
    pdf.set_xy(meta_x, y_start)

    pdf.set_text_color(0, 0, 0)
    pdf.set_font("helvetica", "B", 14)
    pdf.set_text_color(*ACCENT)
    pdf.cell(0, 10, data.title, ln=True)

    pdf.set_font("helvetica", "", 10)
    pdf.set_text_color(85, 85, 85)
    pdf.set_x(meta_x)
    pdf.cell(0, 8, f"Channel: {data.channel_name}", ln=True)
    pdf.set_x(meta_x)
    pdf.cell(0, 8, f"Duration: {data.duration}", ln=True)
    pdf.set_x(meta_x)
    pdf.cell(0, 8, f"Video ID: {data.video_id}", ln=True)

    pdf.ln(15)

    # Summary section
    pdf.set_draw_color(204, 204, 204)
    pdf.line(10, pdf.get_y(), 200, pdf.get_y())
    pdf.ln(5)

    pdf.set_font("helvetica", "B", 12)
    pdf.set_text_color(*ACCENT)
    pdf.cell(0, 10, "Video Summary", ln=True)

    pdf.set_font("helvetica", "", 9)
    pdf.set_text_color(51, 51, 51)
    pdf.multi_cell(0, 6, data.summary)

    pdf.ln(10)

    # Conversation section
    pdf.line(10, pdf.get_y(), 200, pdf.get_y())
    pdf.ln(5)

    pdf.set_font("helvetica", "B", 12)
    pdf.set_text_color(*ACCENT)
    pdf.cell(0, 10, "Conversation", ln=True)
    pdf.ln(5)

    for msg in data.chat_history:
        is_user = msg.role == "user"
        bg_color = USER_BG if is_user else AI_BG
        label = "You" if is_user else "AI Assistant"
        label_color = ACCENT if is_user else (46, 125, 50)  # #2E7D32

        # Check for new page
        if pdf.get_y() > 240:
            pdf.add_page()

        # Message bubble background
        pdf.set_fill_color(*bg_color)
        pdf.rect(10, pdf.get_y(), 190, 25, "F")

        pdf.set_font("helvetica", "B", 8)
        pdf.set_text_color(*label_color)
        pdf.cell(15, 8, label, ln=False)

        pdf.set_font("helvetica", "", 7)
        pdf.set_text_color(136, 136, 136)
        pdf.cell(170, 8, format_timestamp(msg.timestamp), ln=True, align="R")

        pdf.set_font("helvetica", "", 9)
        pdf.set_text_color(34, 34, 34)
        pdf.multi_cell(185, 5, msg.content)

        pdf.ln(8)

    # Optional transcript
    if data.include_transcript and data.transcript:
        pdf.add_page()
        pdf.set_font("helvetica", "B", 14)
        pdf.set_text_color(*ACCENT)
        pdf.cell(0, 10, "Full Transcript", ln=True)
        pdf.ln(5)

        pdf.set_font("helvetica", "", 8)
        pdf.set_text_color(51, 51, 51)
        pdf.multi_cell(0, 4, data.transcript)

    # Output
    pdf_out = pdf.output(dest="S")
    if isinstance(pdf_out, (bytes, bytearray)):
        return bytes(pdf_out)
    return str(pdf_out).encode("latin-1")


async def generate_docx(data: ExportData) -> bytes:
    """Generates a DOCX export and returns it as bytes."""
    doc = Document()

    # Title
    title_para = doc.add_heading(data.title, level=1)

    # Metadata
    doc.add_paragraph(f"Channel: {data.channel_name}")
    doc.add_paragraph(f"Duration: {data.duration}")
    doc.add_paragraph(f"Exported on: {datetime.now().strftime('%Y-%m-%d')}")
    doc.add_paragraph()

    # Summary
    doc.add_heading("Video Summary", level=2)
    doc.add_paragraph(data.summary)
    doc.add_paragraph()

    # Conversation
    doc.add_heading("Conversation", level=2)

    for msg in data.chat_history:
        is_user = msg.role == "user"
        label = "You" if is_user else "AI Assistant"

        if is_user:
            # User messages with shading
            table = doc.add_table(rows=3, cols=1)
            table.style = "Table Grid"
            table.alignment = WD_TABLE_ALIGNMENT.LEFT

            cell = table.rows[0].cells[0]
            shading_elem = OxmlElement("w:shd")
            shading_elem.set(qn("w:fill"), "E3F2FD")
            cell._tc.get_or_add_tcPr().append(shading_elem)

            cell.text = f"{label}\n{msg.content}\n{format_timestamp(msg.timestamp)}"
        else:
            # AI messages as paragraphs
            p = doc.add_paragraph()
            p.add_run(f"{label}\n").bold = True
            p.add_run(msg.content)
            p.add_run(f"\n{format_timestamp(msg.timestamp)}").font.color.rgb = RGBColor(
                136, 136, 136
            )
            doc.add_paragraph()

    # Transcript
    if data.include_transcript and data.transcript:
        doc.add_page_break()
        doc.add_heading("Full Transcript", level=2)
        doc.add_paragraph(data.transcript)

    # Save to buffer
    buffer = io.BytesIO()
    doc.save(buffer)
    buffer.seek(0)
    return buffer.read()


class PDF(FPDF):
    def header(self):
        pass

    def footer(self):
        self.set_y(-15)
        self.set_font("helvetica", "", 8)
        self.set_text_color(136, 136, 136)
        self.cell(0, 10, "YouTube AI Chat Export", align="C")
